### PR TITLE
Replace Download util.

### PR DESCRIPTION
**Pull request by modmuss50**
The current download util is a bit of a mess, and has caused a few issues, this PR replaces what we had with a (hopefully) more robust, testable and maintainable solution.

Features
- eTag support (It seems Mojang's CDN and our meta/maven do not support this right now)
- Age based caching.
- Sha1 based caching.
- HTTP(S) proxy settings should now be respected.
- The downloader has better awareness of offline mode and refresh deps, cleaning up the calling code a bit.
- Uses the new Java 11 HTTP client, provides async support for downloading multiple files
- Progress handling (TODO needs hooking up to gradle)
- Better compression support. Handled by [Methanol](https://mizosoft.github.io/methanol/) 
- Unit tested with a real web server. 

This PR isnt fully complete, just wanted to get it up for inital feedback and let the tests run.

### CODE REVIEW

File: src/main/java/net/fabricmc/loom/api/LoomGradleExtensionAPI.java [hunk 1]
```diff
@@ -56,6 +56,7 @@ public interface LoomGradleExtensionAPI {
 
 	RegularFileProperty getAccessWidenerPath();
 
+	@Deprecated(forRemoval = true)
 	Property<Boolean> getShareRemapCaches();
```
Comment (Juuxel) [inline src/main/java/net/fabricmc/loom/api/LoomGradleExtensionAPI.java:60]: This seems unrelated?

File: src/main/java/net/fabricmc/loom/configuration/providers/mappings/utils/URLFileSpec.java [hunk 2]
```diff
@@ -35,18 +33,20 @@
 
 import net.fabricmc.loom.api.mappings.layered.MappingContext;
 import net.fabricmc.loom.api.mappings.layered.spec.FileSpec;
-import net.fabricmc.loom.util.DownloadUtil;
+import net.fabricmc.loom.util.download.DownloadException;
 
-public record URLFileSpec(URL url) implements FileSpec {
+public record URLFileSpec(String url) implements FileSpec {
```
Comment (Juuxel) [inline src/main/java/net/fabricmc/loom/configuration/providers/mappings/utils/URLFileSpec.java:39]: The hashCode should probably be changed now (or maybe DlBuilder should support URLs?)

File: src/test/groovy/net/fabricmc/loom/test/unit/LoomMocks.groovy [hunk 3]
```diff
@@ -34,12 +37,14 @@ class LoomMocks {
     static IntermediaryMappingsProvider intermediaryMappingsProviderMock(String minecraftVersion, String intermediaryUrl) {
         def minecraftVersionProperty = GradleTestUtil.mockProperty(minecraftVersion)
         def intermediaryUrlProperty = GradleTestUtil.mockProperty(intermediaryUrl)
+		def downloaderProperty = GradleTestUtil.mockProperty(Download.&create as Function)
```
Comment (Juuxel) [inline src/test/groovy/net/fabricmc/loom/test/unit/LoomMocks.groovy]: There's something weird with the indents in this method (mixed tabs and spaces?)

File: src/main/java/net/fabricmc/loom/configuration/providers/mappings/utils/URLFileSpec.java (hunk 2 above)
Comment (modmuss50) [reply src/main/java/net/fabricmc/loom/configuration/providers/mappings/utils/URLFileSpec.java:39]: URL.hashCode [makes a network request](https://stackoverflow.com/questions/2348399/why-does-java-net-urls-hashcode-resolve-the-host-to-an-ip).

I think DLBuilder should also take a URL 👍

File: src/test/groovy/net/fabricmc/loom/test/unit/LoomMocks.groovy (hunk 3 above)
Comment (modmuss50) [reply src/test/groovy/net/fabricmc/loom/test/unit/LoomMocks.groovy]: Humm, looks like we dont check this, will reformat all the groovy to use tabs separately.

File: src/main/java/net/fabricmc/loom/configuration/providers/mappings/utils/URLFileSpec.java (hunk 2 above)
Comment (Juuxel) [reply src/main/java/net/fabricmc/loom/configuration/providers/mappings/utils/URLFileSpec.java:39]: > URL.hashCode [makes a network request](https://stackoverflow.com/questions/2348399/why-does-java-net-urls-hashcode-resolve-the-host-to-an-ip).

Precisely my point, the workaround isn't needed anymore since it's a String now.

**Comment by modmuss50** [comment]
Test failures all seem to be known (caused by previous 0.13 changes) going to merge this now. 👍
